src/s3path.jl

```diff
@@ -355,13 +363,14 @@ function Base.readdir(fp::S3Path; join=false, sort=true)
     end
 end
 
-Base.read(fp::S3Path) = Vector{UInt8}(s3_get(fp.config, fp.bucket, fp.key))
+Base.read(fp::S3Path) = Vector{UInt8}(s3_get(fp.config, fp.bucket, fp.key; version=fp.version))
```

Comment:
Hmmm, I'm starting to wonder if `version` should be part of an `S3Path` or just an argument of a function call. I'll need to think about this.

Reply:
This [particular line](https://github.com/JuliaCloud/AWSS3.jl/pull/161/files#diff-31b593db3c82f33730f4af58846ed59213199a8dd973bddc4f70f1e1721d6719R373) is making me question this design. As the version information is only useful for reading probably we should limit its use to just those kinds functions. I'm unsure on if that will result in higher-level functions not being able to support versioning then

Reply:
Having it as a optional argument feels right to me. But thinking about it a bit, when retrieving a specific version of an object it's part of the URL / path.

> https://{my-bucket}.s3.us-east-1.amazonaws.com/foobar.txt?versionId=BaCWQOBnUvNOjOBm3zY1aTVM0pD8yfdn

Reply:
>This particular line

Yeah---I actually went back and forth on whether the correct behavior there would be to throw an error (or a warning) or just to silently write a new version. I don't have a strong opinion about the correct thing to do.

> As the version information is only useful for reading

Also useful for deleting a specific version! (Although admittedly that is not a use case I've run into much.)

Reply:
> when retrieving a specific version of an object it's part of the URL / path

My use case (and that of others on my team) is such that a large number of the S3 objects that we work with (and there are many!!) _must_ be versioned---the version is a critical part of the object path---which is why I chose this particular implementation.

Until now, I've been using a wonky workaround to call `s3_get` directly for each of these items, e.g.,
```
const dataset_s3path = S3Path("s3://my_bucket/my_key")
const dataset_version = "aweofijwoefjieowjf"

my_dataset = s3_get(dataset_s3path.bucket, dataset_s3path.key; version=dataset_version)
```
This proposed implementation lets us pass around a single S3Path without needing to separately track/pull in a version:
```
const dataset_s3path = S3Path("s3://my_bucket/my_key"; version="aweofijwoefjieowjf")
...
my_dataset = read(dataset_s3path)
```

If `version` is an argument to `read` functions rather than a property of `S3Path` itself, we'll continue to need to separately track versions (and think carefully at all `read` callsites about whether or not to pull them in) for any given file:

```
const dataset_s3path = S3Path("s3://my_bucket/my_key")
const dataset_version = "aweofijwoefjieowjf"
...

my_dataset = read(dataset_s3path; version=dataset_version)
```

...this example is kind of goofy, and makes it seem like I'm averse to keeping track of one extra variable---in reality, we're often keeping track of many paths at the same time, some versioned and some not, which makes the process of additionally tracking versions fairly error-prone, especially when the version truly is a property of the S3 object.

Reply:
I haven't spent too much time thinking about this, I will in the next few days though. I think it does seem like a property of an `S3Path`, and would simplify things in your workflow.

I imagine that currently when working with multiple datasets it just ends up being something like:

```julia
const DATASET_FOO = S3Path("s3:://my_bucket/foo")
const DATASET_FOO_VERSION = "some_identifier"

const DATASET_BAR = S3Path("s3:://my_bucket/bar")
const DATASET_BAR_VERSION = "some_other_identifier"
```

Or a Dictionary, either way that seems bad and like a headache to maintain so I'm all for improving this functionality! I'll take a deeper look at this possibly this afternoon if not tomorrow or early next week.

Reply:
Hah, exactly---that is exactly what it looks like. :) 

Thank you!

Reply:
I've definitely changed my mind and do think `version` makes sense to be part of the `S3Path`. There is some weirdness here but this PR is a step in the right direction.